linkerd/distribute/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+                        .backends
+                        .get_full(keys.next(rng))
+                        .expect("distributions must not reference unknown backends");
+                    if polled_idxs.insert(idx) {
+                        // The index was not already polled, so poll it.
+                        let mut svc = svc.clone();
+                        if svc.poll_ready(cx)?.is_ready() {
+                            self.ready = Some(svc);
+                            return Poll::Ready(Ok(()));
+                        }
+                    }
+                }
+            }
+        }
+
+        debug_assert!(self.ready.is_none());
+        Poll::Pending
+    }
+
+    fn call(&mut self, req: Req) -> Self::Future {
+        self.ready
+            .take()
+            .expect("poll_ready must be called first")
+            .call(req)
+    }
+}
+
+// === impl Distribution ===
+
+impl<K> FromIterator<K> for Distribution<K> {
+    fn from_iter<T: IntoIterator<Item = K>>(iter: T) -> Self {
+        Self::first_available(iter)
+    }
+}
+
+impl<K> FromIterator<(K, u32)> for Distribution<K> {
+    fn from_iter<T: IntoIterator<Item = (K, u32)>>(iter: T) -> Self {
+        Self::random_available(iter)
+    }
+}
+
+impl<K> Distribution<K> {
+    pub fn empty() -> Self {
+        Self::Empty
+    }
+
+    pub fn first_available(keys: impl IntoIterator<Item = K>) -> Self {
+        let keys: Arc<[K]> = keys.into_iter().collect();
+        if keys.is_empty() {
+            return Self::Empty;
+        }
+        Self::FirstAvailable(keys)
+    }
+
+    pub fn random_available<T: IntoIterator<Item = (K, u32)>>(iter: T) -> Self {
+        let (keys, weights): (Vec<_>, Vec<_>) = iter.into_iter().filter(|(_, w)| *w > 0).unzip();
+        if keys.len() < 2 {
+            return Self::first_available(keys);
+        }
+
+        let index = WeightedIndex::new(weights).expect("must succeed");
+        Self::RandomAvailable(Arc::new(WeightedKeys { keys, index }))
+    }
+}
+
+// === impl Inner ===
+
+impl<K> From<Distribution<K>> for Inner<K> {
+    fn from(dist: Distribution<K>) -> Self {
+        match dist {
+            Distribution::Empty => Self::Empty,
+            Distribution::FirstAvailable(keys) => Self::FirstAvailable { keys },
+            Distribution::RandomAvailable(keys) => Self::RandomAvailable {
+                polled_idxs: HashSet::with_capacity(keys.keys.len()),
+                keys,
+                rng: SmallRng::from_rng(rand::thread_rng()).expect("RNG must initialize"),
+            },
+        }
+    }
+}
+
+impl<K> Clone for Inner<K> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Empty => Inner::Empty,
+            Self::FirstAvailable { keys } => Self::FirstAvailable { keys: keys.clone() },
+            Self::RandomAvailable { keys, .. } => Self::RandomAvailable {
+                keys: keys.clone(),
+                polled_idxs: HashSet::with_capacity(keys.keys.len()),
```

Comment:
this could potentially be replaced with a bitset, which could probably avoid allocating in the case where the number of keys is <= 64 and a single integer could be used to represent all the keys. but that's fine to save for a future branch.